src/adapter/src/optimize/index.rs

```diff
@@ -233,10 +233,10 @@ impl<'ctx> Optimize<'ctx, GlobalMirPlan<Resolved>> for OptimizeIndex {
         // Finalize the dataflow. This includes:
         // - MIR ⇒ LIR lowering
         // - LIR ⇒ LIR transforms
+        assert!(!df_desc.is_single_time()); // we are not in a monotonic context here
```

Comment:
I think this should be moved to the `Self::refine_single_time` methods below (see my other comment).